src/normalize/base.ts

```diff
@@ -0,0 +1,41 @@
+import {Config} from '../config';
+import {Encoding} from '../encoding';
+import {Projection} from '../projection';
+import {GenericSpec, NormalizedSpec} from '../spec/index';
+import {GenericLayerSpec, NormalizedLayerSpec} from '../spec/layer';
+import {GenericUnitSpec, NormalizedUnitSpec} from '../spec/unit';
+
+export type Normalize<S extends GenericSpec<any, any>, NS extends NormalizedSpec> = (
+  spec: S,
+  params: NormalizerParams
+) => NS;
+
+export interface ExtraNormalizer<
+  S extends GenericSpec<any, any>, // Input type
+  O extends NormalizedSpec, // Output Type
+  SN extends GenericSpec<any, any> = S // input to additional normalization
+> {
+  name: string;
+  hasMatchingType: (spec: GenericSpec<any, any>, config: Config) => spec is S;
+
+  run(spec: S, params: NormalizerParams, normalize: Normalize<SN, O>): O;
+
+  // TODO: add config cleaner
```

Comment:
Is there an issue for this? What does this even mean?

Reply:
I'll remove this comment